File: cve.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"net/http"
	"os"
	"regexp"
	"time"
)

type CVEFinding struct {
	Package  string
	Version  string
	Severity string
	ID       string
	Summary  string
	URL      string
}

type osvQuery struct {
	Version string     `json:"version"`
	Package osvPackage `json:"package"`
}

type osvPackage struct {
	Name      string `json:"name"`
	Ecosystem string `json:"ecosystem"`
}

type osvResponse struct {
	Results []struct {
		Vulns []struct {
			ID       string `json:"id"`
			Summary  string `json:"summary"`
			Database struct {
				Severity string `json:"severity"`
			} `json:"database_specific"`
		} `json:"vulns"`
	} `json:"results"`
}

var semverClean = regexp.MustCompile(`^[\^~>=<v]`)

func checkCVEs(pkgPath string) ([]CVEFinding, error) {
	data, err := os.ReadFile(pkgPath)
	if err != nil {
		return nil, fmt.Errorf("cannot read %s: %w", pkgPath, err)
	}

	var pkg struct {
		Dependencies    map[string]string `json:"dependencies"`
		DevDependencies map[string]string `json:"devDependencies"`
	}
	if err := json.Unmarshal(data, &pkg); err != nil {
		return nil, fmt.Errorf("invalid package.json: %w", err)
	}

	all := make(map[string]string)
	for k, v := range pkg.Dependencies {
		all[k] = v
	}
	for k, v := range pkg.DevDependencies {
		all[k] = v
	}

	if len(all) == 0 {
		return nil, nil
	}

	// Build queries
	queries := make([]osvQuery, 0, len(all))
	names := make([]string, 0, len(all))
	versions := make([]string, 0, len(all))

	for name, ver := range all {
		clean := semverClean.ReplaceAllString(ver, "")
		queries = append(queries, osvQuery{
			Version: clean,
			Package: osvPackage{Name: name, Ecosystem: "npm"},
		})
		names = append(names, name)
		versions = append(versions, clean)
	}

	// OSV batch API
	body, _ := json.Marshal(map[string]interface{}{"queries": queries})
	client := &http.Client{Timeout: 30 * time.Second}
	resp, err := client.Post("https://api.osv.dev/v1/querybatch", "application/json", bytes.NewReader(body))
	if err != nil {
		return nil, fmt.Errorf("OSV API error: %w", err)
	}
	defer resp.Body.Close()

	var osvResp osvResponse
	if err := json.NewDecoder(resp.Body).Decode(&osvResp); err != nil {
		return nil, fmt.Errorf("OSV response parse error: %w", err)
	}

	var findings []CVEFinding
	for i, result := range osvResp.Results {
		if i >= len(names) {
			break
		}
		for _, vuln := range result.Vulns {
			sev := vuln.Database.Severity
			if sev == "" {
				sev = "UNKNOWN"
			}
			findings = append(findings, CVEFinding{
				Package:  names[i],
				Version:  versions[i],
				Severity: sev,
				ID:       vuln.ID,
				Summary:  vuln.Summary,
				URL:      "https://osv.dev/vulnerability/" + vuln.ID,
			})
		}
	}

	return findings, nil
}

File: src/cve.ts
import { readFileSync } from "fs";

export interface CveFinding {
  package: string;
  version: string;
  ecosystem: string;
  severity: string;
  id: string;
  summary: string;
  url: string;
}

interface OsvQuery {
  version: string;
  package: { name: string; ecosystem: string };
}

interface OsvVuln {
  id: string;
  summary?: string;
  database_specific?: { severity?: string };
}

async function queryOsv(queries: OsvQuery[]): Promise<Map<string, OsvVuln[]>> {
  const res = await fetch("https://api.osv.dev/v1/querybatch", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ queries }),
  });
  if (!res.ok) throw new Error(`OSV API error: ${res.status}`);

  const data = (await res.json()) as { results: Array<{ vulns?: OsvVuln[] }> };
  const map = new Map<string, OsvVuln[]>();

  for (let i = 0; i < queries.length; i++) {
    const key = `${queries[i].package.name}@${queries[i].version}`;
    map.set(key, data.results[i]?.vulns ?? []);
  }

  return map;
}

function parseDeps(pkgPath: string): Array<{ name: string; version: string }> {
  let raw: string;
  try {
    raw = readFileSync(pkgPath, "utf-8");
  } catch {
    throw new Error(`Cannot read file: ${pkgPath}`);
  }

  const pkg = JSON.parse(raw) as {
    dependencies?: Record<string, string>;
    devDependencies?: Record<string, string>;
  };

  const all = {
    ...pkg.dependencies,
    ...pkg.devDependencies,
  };

  return Object.entries(all).map(([name, version]) => ({
    name,
    version: version.replace(/^[\^~>=<]/, ""),
  }));
}

export async function checkCves(pkgJsonPath: string): Promise<CveFinding[]> {
  const deps = parseDeps(pkgJsonPath);
  if (deps.length === 0) return [];

  const BATCH = 100;
  const findings: CveFinding[] = [];

  for (let i = 0; i < deps.length; i += BATCH) {
    const chunk = deps.slice(i, i + BATCH);
    const queries: OsvQuery[] = chunk.map((d) => ({
      version: d.version,
      package: { name: d.name, ecosystem: "npm" },
    }));

    const results = await queryOsv(queries);

    for (const [key, vulns] of results) {
      if (!vulns.length) continue;
      const [name, version] = key.split("@");
      for (const vuln of vulns) {
        findings.push({
          package: name,
          version,
          ecosystem: "npm",
          severity: vuln.database_specific?.severity ?? "UNKNOWN",
          id: vuln.id,
          summary: vuln.summary ?? "No summary available",
          url: `https://osv.dev/vulnerability/${vuln.id}`,
        });
      }
    }
  }

  return findings;
}
